### PR TITLE
Reduce likelyhood of race in EventMap

### PR DESCRIPTION
**Pull request by mattwar**
Possible fix for TFS devdiv bug 1154956

This change makes it so that in-flight events that have not yet fired (due to asynchronous scheduling), will not be fired on handlers that have unsubscribed from the event.  

It is still possible that an event handler can be entered after it has been unsubscribed from a thread that is different than the one that fires the event.

@Pilchie please take a look

### CODE REVIEW

File: src/Workspaces/Core/Portable/Utilities/EventMap.cs [hunk 1]
```diff
@@ -18,64 +19,133 @@ public EventMap()
...
         public void AddEventHandler<TEventHandler>(string eventName, TEventHandler eventHandler)
+            where TEventHandler : class
         {
             using (_guard.DisposableWait())
             {
-                var handlers = GetEvents_NoLock<TEventHandler>(eventName);
-                var newHandlers = handlers.Add(eventHandler);
-                SetEvents_NoLock(eventName, newHandlers);
+                var registries = GetRegistries_NoLock<TEventHandler>(eventName);
+                var newHandlers = registries.Add(new Registry<TEventHandler>(eventHandler));
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Utilities/EventMap.cs]: rename to newRegistries?

File: src/Workspaces/Core/Portable/Utilities/EventMap.cs [hunk 2]
```diff
@@ -18,64 +19,133 @@ public EventMap()
...
+
+                // remove disabled registrations from list
+                var newRegistries = registries.RemoveAll(r => r.HasHandler(eventHandler));
+
+                if (newRegistries != registries)
                 {
-                    SetEvents_NoLock(eventName, newHandlers);
+                    // disable all registrations of this handler (so pending raise events can be squelched)
+                    // This does not guarantee no race condition between Raise and Remove but greatly reduces it.
+                    foreach (var registery in registries.Where(r => r.HasHandler(eventHandler)))
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Utilities/EventMap.cs]: spelling - no second e in registry?

File: src/Workspaces/Core/Portable/Utilities/EventMap.cs [hunk 3]
```diff
@@ -18,64 +19,133 @@ public EventMap()
...
+                var registries = GetRegistries_NoLock<TEventHandler>(eventName);
+
+                // remove disabled registrations from list
+                var newRegistries = registries.RemoveAll(r => r.HasHandler(eventHandler));
+
+                if (newRegistries != registries)
                 {
-                    SetEvents_NoLock(eventName, newHandlers);
+                    // disable all registrations of this handler (so pending raise events can be squelched)
+                    // This does not guarantee no race condition between Raise and Remove but greatly reduces it.
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Utilities/EventMap.cs]: There is always a race for multi-threaded apps, but this closes the race completely for apps with an STA because both the Remove and the Raise will happen on that same thread through the synchronization context?

File: src/EditorFeatures/CSharpTest/Workspaces/WorkspaceTests.cs [hunk 4]
```diff
@@ -48,7 +48,7 @@ public void TestEmptySolutionUpdate()
                 WaitForWorkspaceOperationsToComplete(workspace);
 
                 Assert.Equal(solution, workspace.CurrentSolution);
-                Assert.False(workspaceChanged);
+                Assert.True(workspaceChanged);
```
Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/Workspaces/WorkspaceTests.cs]: I don't understand why this changed.

Comment (mattwar) [inline src/EditorFeatures/CSharpTest/Workspaces/WorkspaceTests.cs]: Me either

Comment (mattwar) [inline src/EditorFeatures/CSharpTest/Workspaces/WorkspaceTests.cs]: It obviously needs to be assert true, because the whole purpose of the test is to prove the event is fired.  I don't know why it was the other way, or why the change makes the event become observable now and not before.

File: src/Workspaces/Core/Portable/Utilities/EventMap.cs (hunk 3 above)
Comment (mattwar) [inline src/Workspaces/Core/Portable/Utilities/EventMap.cs]: Yes.  And the indirection through the registry type reduces any remaining races further (given separate threads for event and remove).

**Comment by Pilchie** [comment]
:+1:

**Comment by sharwell** [comment]
I'm... "not thrilled about this." However, with limited information about the specific bug you are trying to solve, I can only give subjective feedback.

* I don't like terminology like "greatly reduces it" when talking about the chance for race conditions. I prefer to see accurate descriptions of the limitations of the algorithm. For example:

    > Disable all registrations of this handler. This will prevent the handler from getting called unless a concurrent operation on another thread is already in the process of invoking the handler.

* `HasHandler` is a misleading name. `TEventHandler` is a multicast delegate instance, so technically the input to this method could be a single call within the invocation list of the multicast delegate. However, this is not what the method is looking for. I think the most straightforward implementation would be creating a property `Handler` and using `r.Handler == eventHandler`.

* Well-known concurrent algorithms could eliminate race conditions provided the concurrent behavior is documented in the class along with information about expected usage patterns. None of this information is provided. Of course this isn't a problem introduced by this pull request, but it appears you are facing concurrency problems with a type already meeting and I :frowning: to see band-aids but limited assurance of a solution.

**Comment by Pilchie** [comment]
We should probably document this a little more clearly, however, I don't think it's quite as bad as you fear. 

In VS, (or any app where the `Workspace` is initialized on an STA thread), there is no race after @mattwar's fix, because both the unsubscribe and the raise happen on the STA thread, and the raise re-checks the list of subscribed handlers inside the task).

There is still a possibility of an event being fired after unsubscribe if the unsubscribe happens on a different thread, but that is no different than any .NET event.

The one place where there is still a small race is in something like a console app.  Because in that case the events will fire on the threadpool, but the unsubscribe takes place on the original thread, there is a slight chance that the event will still be raised after the unsubscribe.

**Comment by heejaechang** [comment]
@Pilchie @mattwar I am not sure whether subscribing and unsubscribing actually happen on UI thread. I don't see, AddEventHandler/RemoveEventHandler switching thread to UI thread. 

but regardless of that, it basically get list of event handlers at the point of raising events, not at the point of events happened. and use that list rather than getting a snapshot of listener at the point of events happen and use it.

I can understand why it is using Registry type to remove possibility of dead lock, but with per handler lock + the way it works, that it needs to hold while raising event for the handler, and that needs to be checked per while removing/adding handler, we might be able to remove race completely regardless of thread.

if we can remove race regardless of thread, then, I think we can move events to BG.

**Comment by heejaechang** [comment]
I should have said, by per handler lock, I meant re-entrance lock which let same thread to go in, and block different thread to blocked.

also if we have HasEventHandler and RaiseEvent as two different calls, won't be a race? two seems must be under same lock.

**Comment by mattwar** [comment]
@heejaechang You cannot remove race completely without using a lock per event handler and blocking the remove/unsubscribe calls while events are in flight.  And you never, ever, want to run random user code under a lock.  So we are left with being generally okay with the idea that in some circumstances event handlers may be invoked after they have been unsubscribed if the event was already in flight.
